### ISSUE TITLE
Volatile is deprecated in C++20

### ISSUE THREAD

**Issue by Troels51** [open-source-parsers/jsoncpp#1193]
**Describe the bug**
In C++20 a lot of the volatile keyword is being deprecated ([P1152R3](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1152r3.html))
For example [deallocate](https://github.com/open-source-parsers/jsoncpp/blob/632044ad956b9ca55703d98d852bc815077f6afc/include/json/allocator.h#L41) has this problem

**To Reproduce**
1. Use a new compiler, fx gcc10
2. Set CMAKE_CXX_STANDARD to 20
3. Build

**Expected behavior**
Expected the build to complete

**Desktop (please complete the following information):**
- gcc version 10.1.0

**Additional context**
I ran into this problem because i am using jsoncpp in a project that was setting CXX_STANDARD to 20, and then when we upgraded gcc we were getting compile errors.
For example
```
jsoncpp-src/include/json/allocator.h:41:36: error: ‘volatile’-qualified parameter is deprecated [-Werror=volatile]
   41 |   void deallocate(volatile pointer p, size_type n) {
      |                   ~~~~~~~~~~~~~~~~~^
```

**Comment by dota17** [comment]
I think we are far from ready to adapt to the C + + 20 standard. But if you like, you can submit a PR to fix this, and I'd love to review it then.

**Comment by cdunn2001** [comment]
@dawesc, any opinion? This was originally your code, apparently.

https://en.cppreference.com/w/cpp/language/cv

> Some uses of volatile are deprecated: ...
>
> volatile object type as function parameter type or return type;

**Comment by ledergec** [comment]
This seems to be the only reason which prevent to compile with c++20 standard using clang 10. The following patch fixes the issue:

diff --git a/include/json/allocator.h b/include/json/allocator.h
index d5f987e..b6c1fdf 100644
--- a/include/json/allocator.h
+++ b/include/json/allocator.h
@@ -35,11 +35,12 @@ public:
    * Release memory which was allocated for N items at pointer P.
    *
    * The memory block is filled with zeroes before being released.
-   * The pointer argument is tagged as "volatile" to prevent the
-   * compiler optimizing out this critical step.
+   * The pointer argument passed to memset is tagged as "volatile" to prevent
+   * the compiler optimizing out this critical step.
    */
-  void deallocate(volatile pointer p, size_type n) {
-    std::memset(p, 0, n * sizeof(T));
+  void deallocate(pointer p, size_type n) {
+    volatile pointer do_not_optimize_memset_p = p;
+    std::memset(do_not_optimize_memset_p, 0, n * sizeof(T));
     // free using "global operator delete"
     ::operator delete(p);
   }

**Comment by ledergec** [comment]
I have create the following PR to fix this:
https://github.com/open-source-parsers/jsoncpp/pull/1228

**Comment by BillyDonahue** [comment]
Fixed by #1228
